var/spack/repos/builtin/packages/graphblast/package.py

```diff
@@ -0,0 +1,42 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+
+
+class Graphblast(MakefilePackage, CudaPackage):
+    """High-Performance Linear Algebra-based Graph Primitives on GPUs"""
+
+    homepage    = "https://github.com/gunrock/graphblast"
+    git         = "https://github.com/gunrock/graphblast.git"
+
+    version('master', submodules=True)
+
+    variant('cuda', default=True)
+
+    depends_on('boost +program_options')
+
+    # This package is confirmed to compile with gcc <= 5.4.0,boost@1.58,cuda@9
+    # TODO: the package doesn't compile as CMakePackage currently
+    # once that is fixed it should be converted to a CMakePackage type.
+
+    conflicts('%gcc@5.5.0:')
+
+    conflicts('cuda_arch=none', when='+cuda',
+              msg='Must specify CUDA compute capabilities of your GPU. \
+See "spack info graphblast"')
+
+    def install(self, spec, prefix):
+        install_tree(self.build_directory, self.prefix)
+
+    @run_before('build')
+    def set_cudarch(self):
```

Comment:
Maybe do this in `patch`?

Reply:
Will do. I sometimes forget about that and do file filtering which is more work!

Reply:
Actually writing a patch is not an option since we don't know to begin with Cuda arch user wants to compile this software for. May be `cudapackage` class has a way to do it, and hence we would disable this within the Makefile.

Reply:
I meant a `patch(self)` method.

Reply:
Does that run before build phase since the documentation is not clear about this and it just says it will run before `install()` phase.

Reply:
I tried it:
```diff
diff --git a/var/spack/repos/builtin/packages/postgis/package.py b/var/spack/repos/builtin/packages/postgis/package.py
index 719b36c..0153355 100644
--- a/var/spack/repos/builtin/packages/postgis/package.py
+++ b/var/spack/repos/builtin/packages/postgis/package.py
@@ -71,8 +71,7 @@ def install(self, spec, prefix):
              'pkglibdir=' + prefix.lib, 'datadir=' + prefix.share,
              'docdir=' + prefix.share.doc)
 
-    @run_before('build')
-    def fix_raster_bindir(self):
+    def patch(self):
         makefile = FileFilter('raster/loader/Makefile')
         makefile.filter('$(DESTDIR)$(PGSQL_BINDIR)', self.prefix.bin, string=True)
         makefile = FileFilter('raster/scripts/Makefile')
```
but I think I am missing something:
```
==> Created stage in /cache/sbulut/spack-stage/spack-stage-postgis-3.0.1-xb4amzj7fvkp7ghjbzzysspepaspltr4
==> Error: FileNotFoundError: [Errno 2] No such file or directory: 'raster/loader/Makefile'

/mnt/local/spack/var/spack/repos/builtin/packages/postgis/package.py:76, in patch:
         74    def patch(self):
         75        makefile = FileFilter('raster/loader/Makefile')
  >>     76        makefile.filter('$(DESTDIR)$(PGSQL_BINDIR)', self.prefix.bin, string=True)
         77        makefile = FileFilter('raster/scripts/Makefile')
         78        makefile.filter('$(DESTDIR)$(PGSQL_BINDIR)', self.prefix.bin, string=True)
```

Reply:
It runs before all other phases